### DISCUSSION TITLE
Refactoring the easydiffraction library

### DISCUSSION THREAD

**Discussion by AndrewSazonov** [EasyScience/EasyDiffractionLib#156] [Ideas]
In order to simplify code maintenance and following the ideas for working with calculators in EasyReflectometry, the following changes to the project structure are suggested.

## Step 1

* easydiffraction - Move to src/easydiffraction
  * calculators - As before
    * \_\_init\_\_.py - As before
    * CFML.py - Move to src/easydiffraction/calculators/pycrysfml/calculator.py
      * class CFML - Rename to Pycrysfml
    * cryspy.py - Move to src/easydiffraction/calculators/cryspy/calculator.py
      * class Cryspy - As before
    * GSASII.py - Remove
    * pdffit2.py - Move to src/easydiffraction/calculators/pdffit2/calculator.py
      * class Pdffit2 - As before
  * components - Temporary as before (to be addressed in Step 2)
    * \_\_init\_\_.py
    * phase.py
    * polarization.py
    * site.py
  * elements - Temporary as before (to be addressed in Step 2)
    * Backgrounds
      * \_\_init\_\_.py
      * Background.py
      * Factorial.py
      * Point.py
    * Experiments
      * \_\_init\_\_.py
      * Experiment.py
      * Pattern.py
    * \_\_init\_\_.py
  * Interfaces - Remove
    * \_\_init\_\_.py - Add to src/easydiffraction/calculators/\_\_init\_\_.py
    * CFML.py - Move to src/easydiffraction/calculators/pycrysfml/wrapper.py
      * class CFML - Rename to PycrysfmlWrapper
    * cryspy.py - Remove
    * cryspyV2.py - Move to src/easydiffraction/calculators/cryspy/wrapper.py
      * class CryspyV2 - Rename to CryspyWrapper
    * GSASII.py - Remove
    * interfaceTemplate.py - Move to src/easydiffraction/calculators/wrapper_base.py
      * class InterfaceTemplate - Rename to WrapperBase
    * pdffit2.py - Move to src/easydiffraction/calculators/pdffit2/wrapper.py
      * class Pdffit2 - Rename to Pdffit2Wrapper
    * types.py - Move to src/easydiffraction/calculators/wrapper_types.py (???)
  * io - Remove
    * \_\_init\_\_.py - Remove
    * cif.py - Remove  
      * cifV2ToV1 - Move to src/easydiffraction/calculators/cryspy/parser.py
      * cifV2ToV1_tof - Move to src/easydiffraction/calculators/cryspy/parser.py
    * cif_reader.py - From EDB. Move to EasyCrystallography
    * cryspy_parser.py - Move to src/easydiffraction/calculators/cryspy/parser.py
    * download.py - Remove
      * def download_from_repository -> Move to src/easydiffraction/utils.py
    * helpers.py - Remove
  * Profiles - Temporary as before (to be addressed in Step 2)
    * \_\_init\_\_.py
    * Analysis.py
    * common.py
    * Container.py
    * Experiment.py
    * JobType.py
    * P1D.py
    * Sample.py
  * \_\_init\_\_.py - As before
  * interfaces.py - Move to src/easydiffraction/calculators/wrapper_factory.py
    * class InterfaceFactory - Rename to WrapperFactory
  * Job.py - Rename to job.py
  * Jobs.py - Remove
  * main.py - As before
  * Runner.py - Remove
  * sample.py - Temporary as before (to be addressed in Step 2)
  
  The next steps will follow when the first one is implemented.

**Comment by rozyczko** [comment]
> cif.py - Remove

Also, `dataBlockToCif` is referenced in the application, so it should not be removed

The remaining classes `CifIO` etc. should probably be retained for potential future use as the content seems very useful.

**Reply by AndrewSazonov** [reply]
At this point, I've added `cif.py` back. Let's take a closer look at their functionality when the main refactoring is complete.

**Comment by rozyczko** [comment]
> cryspy_parser.py

Only `class Parameter` and method  `def calcObjAndDictToEdExperiments` need to be retained. The method can be removed from the class and referenced as a static, standalone method.

**Reply by AndrewSazonov** [reply]
As I see `cifV2ToV1_tof` and `cifV2ToV1` are also needed...

**Reply by rozyczko** [reply]
I mean, yes, apart of these two , which we discussed earlier :)

**Comment by rozyczko** [comment]
`cif_reader.py` probably should not be moved to Crystallography.
It takes CIF blocks and converts them into diffraction-specific dicts. Let's just keep it here.

**Comment by AndrewSazonov** [comment]
After Step 1, the structure is as follows:

* src
  * easydiffraction
    * calculators
      * cryspy
        * calculator.py
          * class Cryspy
        * parser.py
        * wrapper.py
          * class CryspyWrapper
      * pdffit2
        * calculator.py
          * class Pdffit2
        * wrapper.py
          * class Pdffit2Wrapper
      * pycrysfml
        * calculator.py
          * class Pycrysfml
        * wrapper.py
          * class PycrysfmlWrapper
      * \_\_init\_\_.py
      * wrapper_base.py
        * class WrapperBase
      * wrapper_factory.py
        * class WrapperFactory
      * wrapper_types.py
    * components
      * \_\_init\_\_.py
      * phase.py
      * polarization.py
      * site.py
    * elements
      * Backgrounds
        * \_\_init\_\_.py
        * Background.py
        * Factorial.py
        * Point.py
      * Experiments
        * \_\_init\_\_.py
        * Experiment.py
        * Pattern.py
      * \_\_init\_\_.py
    * io
      * \_\_init\_\_.py
      * cif.py
      * cif_reader.py
    * Profiles
      * \_\_init\_\_.py
      * Analysis.py
      * common.py
      * Container.py
      * Experiment.py
      * JobType.py
      * P1D.py
      * Sample.py
    * \_\_init\_\_.py
    * job.py
    * main.py
    * sample.py
    * utils.py

**Comment by rozyczko** [comment]
For Step 2, starting with which files should be retained/deleted

## Step 2

  * Components
    * \_\_init\_\_.py
    * phase.py - stays
    * polarization.py - currently unused but will, when adding polarized
    * site.py - unused, but might contain useful classes
  * elements
    * Backgrounds
      * \_\_init\_\_.py
      * Background.py - stays
      * Factorial.py  - unused, but might be useful later, do not delete
      * Point.py - stays
    * Experiments
      * \_\_init\_\_.py
      * Experiment.py - unused, delete
      * Pattern.py  - unused, delete
    * \_\_init\_\_.py
 * Profiles
   *  Analysis.py - stays
   * common.py - stays
     * class PowderBase - remove
   * Container.py - stays
   * Experiment.py - stays
   * JobType.py - stays
   * P1D.py - stays
   * Sample.py - stays

**Comment by AndrewSazonov** [comment]
## Suggested changes in step 2

* src
  * easydiffraction
    * ...
    * components - Remove
      * \_\_init\_\_.py - Remove
      * phase.py - Move to src/easydiffraction/job/model/phase.py (Check if we can use the EC phase directly)
      * polarization.py - Move to src/easydiffraction/job/experiment/polarization.py
      * site.py - Move to src/easydiffraction/job/model/site.py (Check if we can use the EC phase directly)
        * class PeriodicSite - Remove 
        * class PeriodicAtoms - Remove 
    * elements - Remove
      * Backgrounds - Move to src/easydiffraction/job/experiment/backgrounds
        * \_\_init\_\_.py - Move to src/easydiffraction/job/experiment/backgrounds/\_\_init\_\_.py
        * Background.py - Move to src/easydiffraction/job/experiment/backgrounds/background.py
        * Factorial.py - Move to src/easydiffraction/job/experiment/backgrounds/factorial.py
        * Point.py - Move to src/easydiffraction/job/experiment/backgrounds/point.py
      * Experiments - Remove
        * \_\_init\_\_.py - Remove
        * Experiment.py - Remove
        * Pattern.py - Remove
      * \_\_init\_\_.py - Remove
    * ...
    * Profiles
      * \_\_init\_\_.py - Remove
      * Analysis.py - Move to src/easydiffraction/job/analysis/analysis.py
      * common.py - Move to src/easydiffraction/job/experiment/common.py
        * class _PowderBase - Remove
      * Container.py - Move to src/easydiffraction/job/experiment/data_container.py
      * Experiment.py - Move to src/easydiffraction/job/experiment/experiment.py
      * JobType.py - Move to src/easydiffraction/job/experiment/experiment_type.py
        * class JobType - Rename to ExperimentType
      * P1D.py - Move to src/easydiffraction/job/experiment/pd_1d.py
      * Sample.py - Check if we can get rid of this file. If any of these methods are used, can we move them to another location?
    * \_\_init\_\_.py
    * job.py - Move to src/easydiffraction/job/job.py
    * main.py
    * sample.py - It behaves as a layer between Job and the main components of Job (model, experiment, ...). Check if we can get rid of this file and move its functionality to other classes. If this cannot be done quickly, move this file temporarily, e.g. to old_sample.py
    * utils.py


## Expected structure after step 2

* src
  * easydiffraction
    * calculators
      * cryspy
        * calculator.py
          * class Cryspy
        * parser.py
        * wrapper.py
          * class CryspyWrapper
      * pdffit2
        * calculator.py
          * class Pdffit2
        * wrapper.py
          * class Pdffit2Wrapper
      * pycrysfml
        * calculator.py
          * class Pycrysfml
        * wrapper.py
          * class PycrysfmlWrapper
      * \_\_init\_\_.py
      * wrapper_base.py
        * class WrapperBase
      * wrapper_factory.py
        * class WrapperFactory
      * wrapper_types.py
    * job
      * analysis
        * \_\_init\_\_.py
        * analysis.py
      * experiment
        * backgrounds
          * \_\_init\_\_.py
          * background.py
          * factorial.py
          * point.py
        * \_\_init\_\_.py
        * common.py 
        * data_container.py
        * experiment.py
        * experiment_type.py
          * class ExperimentType 
        * old_sample.py?
        * pd_1d.py 
        * polarization.py 
      * model
        * \_\_init\_\_.py
        * phase.py  
        * site.py
      * project
        * \_\_init\_\_.py
      * summary
        * \_\_init\_\_.py
      * \_\_init\_\_.py
      * job.py
    * \_\_init\_\_.py
    * main.py
    * utils.py

**Comment by rozyczko** [comment]
looks good but I would put `old_sample` in the `model` subdirectory, not experiment. 
It's a container for (among other things) `Phase` and `Phases`, so probably belongs there better.

**Reply by AndrewSazonov** [reply]
- `src/easydiffraction/sample.py` moved to `src/easydiffraction/job/old_sample/old_sample.py`
- `src/easydiffraction/Profiles/Sample.py` moved to `src/easydiffraction/job/experiment/simulation.py`
